### ISSUE TITLE
Using Update command with table-per-hierarchy class does not set the discriminator 

### ISSUE THREAD

**Issue by JonPSmith** [dotnet/efcore#7510]
If I use the `Update` command to change the type of an entry in a table-per-hierarchy class, i.e. change the class from `Book` to `BookWithPromotion` then the discriminator column is not updated. 

---

~~I did try setting the discriminator in code, but that didn't work either~~
## Update
@ajcvickers pointed out that I needed to set the discriminator to readwrite using the following code in my `OnModelCreating` class
```c#
modelBuilder.Entity<Book>()
    //EF Core version 1.x syntax
    .Property("HasPromotion").Metadata.IsReadOnlyAfterSave = false; 
    //EF Core version 2.0.0 syntax
    .Property("HasPromotion").Metadata.AfterSaveBehavior = PropertySaveBehavior.Save 
```
Then the code below works.

---

### Steps to reproduce
The following Unit Test fails both tests at the end. 
```c#
[Fact]
public void UpdateBookToPromotionNoTracking()
{
    //SETUP
    const string connection =
         "Data Source=(localdb)\\mssqllocaldb;Database=TestThp;Integrated Security=True;";       
    var optionsBuilder =  new DbContextOptionsBuilder <MyDbContext>();             

    var options = optionsBuilder.UseSqlServer(connection).Options;

    int bookId;
    using (var context = new MyDbContext(options))
    {
        context.Database.EnsureCreated();

        var book = new Book {Title = "Book", Price = 2};
        context.Books.Add(book);
        context.SaveChanges();
        bookId = book.BookId;
    }

    using (var context = new MyDbContext(options))
    {
        //ATTEMPT
        var bookNoTracking = context.Books.AsNoTracking().Single(x => x.BookId == bookId);

        var bookProm = new BookPromotion
        {
            BookId = bookNoTracking.BookId,
            Title = bookNoTracking.Title,
            OldPrice = bookNoTracking.Price,
            //promotion stuff
            HasPromotion = true,
            Price = 1,
            PromotionText = "This promotion"
        };
        context.Books.Update(bookProm);
        context.SaveChanges();
    }
    using (var context = new MyDbContext(options))
    {
        //VERIFY
        context.Books.Single(x => x.BookId == bookId).GetType().ShouldEqual(typeof(BookPromotion));
        context.Books.Single(x => x.BookId == bookId).HasPromotion.ShouldBeTrue();
    }
}
```
The classes are:
```c#
    public class Book
    {
        public int BookId { get; set; }
        public bool HasPromotion { get; set; }
        public string Title { get; set; }
        public decimal Price { get; set; }
    }
```
```c#
    public class BookPromotion : Book
    {
        public decimal OldPrice { get; set; }
        public string PromotionText { get; set; }
    }
```
The DbContext is:
```c#
public class MyDbContext : DbContext
{
    public DbSet<Book> Books { get; set; }

    public MyDbContext(
        DbContextOptions<MyDbContext> options)      
        : base(options) { }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Book>()
            .HasDiscriminator<bool>("HasPromotion")
            .HasValue<Book>(false)
            .HasValue<BookPromotion>(true);
    }
}
```

### Further technical details
EF Core version: 1.1.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2015, Update 3.

**Comment by ajcvickers** [comment]
Note for triage: The reason for this is that shadow properties are marked as having an unknown value when an entity is attached to the context. This stops accidental over writing with null of, for example, FK properties--see #4920. In this case it means that the discriminator property is the shadow property, and the value it should have is known because it comes from the CLR type. So we could special case discriminators, although off the top of my head I'm not sure where the special casing would go given that discriminators are a relational concept.

Also note that this is an attempt to do back-door type mutation, which is why the code is jumping through all the hoops of creating a no-tracking query, etc. So we could just say that this is not supported.

**Comment by JonPSmith** [comment]
Hi @ajcvickers,

OK, I can see it could be a lot of work to get Update to work with discriminators. If that is the decision I will put that in my write up.

**Comment by ajcvickers** [comment]
We discussed this in triage and decided that we won't add explicit support for mutating the type in this way. If the discriminator does need to be set in an update (only needed to mutate the type) then the discriminator value can be set explicitly.

**Comment by JonPSmith** [comment]
HI @ajcvickers,

Fine, but re your last comment:

> If the discriminator does need to be set in an update (only needed to mutate the type) then the discriminator value can be set explicitly.

I did try that and setting the discriminator directly **did not work** (I mentioned that in my initial post - see second paragraph in italics).

**Comment by ajcvickers** [comment]
@JonPSmith It works for me if I set the property to allow writes:
```C#
modelBuilder.Entity<Book>()
    .Property("Discriminator").Metadata.IsReadOnlyAfterSave = false;
```
And then set the value before saving:
```C#
context.Entry(bookProm).Property("Discriminator").CurrentValue = nameof(BookPromotion)
```

What error are you seeing?

**Comment by JonPSmith** [comment]
Hi @ajcvickers,

Well, another learning experience. I hadn't got into the `.MetaData` part of setting up properties - thanks for pointing that out! I need to pull that into my examples.

Yes, when I added the `...Metadata.IsReadOnlyAfterSave = false;` to my Fluent API my unit test (with `HasPromotion = true,` added) worked. Previously, because it was read only, setting it didn't work.

I have updated my original post so that if anyone else has this problem then they can see the answer
